docs/design/features/MarshalDirective.md

```diff
@@ -0,0 +1,206 @@
+
+    * This process is also influenced by additional `DllImportAttribute` properties as well as by the underlying platform. For example, the [Win32 API ANSI/UNICODE convention](https://docs.microsoft.com/windows/win32/intl/conventions-for-function-prototypes) on Windows is respected and a(n) `A`/`W` suffix may be appended to the function name if it is not immediately found.
+
+4) The IL Stub is called like any .NET method, but the address of the export is passed to the generated IL Stub via a 'hidden' argument.
+
+    * **Marshal Directive**: The user's defined method would require an identical signature to that of the one typically defined, but a new argument of type `IntPtr` should be added as the first argument. The runtime would be responsible for providing the export to the user defined function.
+
+5) The IL Stub then marshals arguments as appropriate and invokes the export via the `calli` instruction.
+
+    * **Marshal Directive**: The user's defined method would cast the supplied export to an appropriate C# function pointer type and invoke the call.
+
+6) Once the export returns control to the IL Stub the marshaling logic cleans up and ensures any returned data is marshaled back out to the calling function.
+
+An example of how the snippet above could be transformed is below. The example is using the proposed API in this document as well as the [C# function pointer proposal][csharp_fptr_link]. Observe points (C), the new attribute, (D), a `partial` class, and (E) the stub with additional `IntPtr` argument that will be called in lieu of a runtime generated IL Stub.
+
+``` CSharp
+[DllImportAttribute("Kernel32.dll")]
+/* C */ [MarshalDirectiveAttribute(
+        Type = MarshalDirectiveType.UserDefinedStub,
+        UserFunctionType = typeof(Stubs),
+        UserDefinedStub = nameof(QueryPerformanceCounterStub))]
+/* A */ extern static bool QueryPerformanceCounter(out long lpPerformanceCount);
+...
+long count;
+/* B */ QueryPerformanceCounter(out count);
+
+/* D */  class partial Stubs
+{
+    /* E */ public static bool QueryPerformanceCounterStub(IntPtr fptr, out long lpPerformanceCount)
+    {
+        unsafe
+        {
+            long result = 0;
+            var tfptr = (delegate* stdcall<long*, int>)fptr;
+            bool success = tfptr(&result) != 0;
+            lpPerformanceCount = result;
+            return success;
+        }
+    }
+}
+```
+
+The contents of the above stub are trivial and in a `Release` build wouldn't even exist. More complicated examples may require [GC handle pinning](https://docs.microsoft.com/dotnet/api/system.runtime.interopservices.gchandle) or native memory allocation. These complicated examples could be hand authored or generated through a third-party tool.
```

Comment:
I don't understand the comment about `Release` builds. I thought that one of the pros of this design is that it allows me to write a custom marshaller for native methods with weird signatures. So the logic I write into the custom stub is non-trivial and runtime won't be able to emulate it. Consequently it can't be different between Debug/Release as runtime would not know what to do without it.

Reply:
That was for the example. Take another look and see if this issue still applies.